src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -1081,11 +1187,77 @@ SDCPUThread::recordCycleStats()
 }
 
 void
-SDCPUThread::squashUpTo(shared_ptr<InflightInst> inst_ptr, bool rebuild_map)
+SDCPUThread::sendToMemory(weak_ptr<InflightInst> inst,
+                          const RequestPtr& req, bool write,
+                          shared_ptr<uint8_t> data,
+                          shared_ptr<SplitRequest> sreq)
+{
+    const shared_ptr<InflightInst> inst_ptr = inst.lock();
+    if (!inst_ptr || inst_ptr->isSquashed()) {
+        // No need to do anything for an instruction that has been squashed.
+        return;
+    }
+
+    sendToMemory(inst_ptr, req, write, data, sreq);
+}
+
+void
+SDCPUThread::sendToMemory(shared_ptr<InflightInst> inst_ptr,
```

Comment:
I'm a little surprised this compiles. I thought that `shared_ptr` could be implicitly cast to `weak_ptr`. How does the compiler know which one to call? Even if the compiler *can* infer the right function, It's confusing for people reading the code.

Reply:
Well it's implicitly castable but it's still a different type... the compiler will always try to match exactly first, then try any casts, so if there were two things that a shared_ptr could be casted to, then it would probably complain.

This was something I thought to do after you showed me the perf details because there are cases where I call sendToMemory with a weak_ptr and with a shared_ptr. This overload allows us to make both calls safely, while minimizing the number of extra weak_ptrs being constructed, and the number of shared_ptr being created from .lock().